feature/experimental/p4rt/tests/p4rt_election/p4rt_election_test.go

```diff
@@ -0,0 +1,617 @@
+	})
+	if err != nil {
+		t.Logf("Write Error: %v", err)
+		return err
+	}
+	return nil
+}
+
+// Function to read a table entry returns response/error
+func readTableEntry(args *testArgs, t *testing.T) (*p4_v1.ReadResponse, error) {
+	t.Helper()
+	readMsg := &p4_v1.ReadRequest{
+		DeviceId: args.deviceID,
+		Entities: []*p4_v1.Entity{
+			{
+				Entity: &p4_v1.Entity_TableEntry{
+					TableEntry: &p4_v1.TableEntry{
+						TableId: p4rtutils.WbbTableMap["acl_wbb_ingress_table"],
+					},
+				},
+			},
+		},
+	}
+	readClient, err := args.handle.Read(readMsg)
+	if err != nil {
+		t.Logf("ReadClient Error: %v", err)
+		return nil, err
+	}
+	readResp, err := readClient.Recv()
+	if err != nil {
+		t.Logf("readResp Error: %v", err)
+		return nil, err
+	}
+	return readResp, nil
+}
+
+// Every client/controller should be able to read
+func canRead(t *testing.T, args *testArgs) (bool, error) {
+	_, readErr := readTableEntry(args, t)
+	if readErr != nil {
+		return false, readErr
+	}
+	return true, nil
+}
+
+// Only Primary should be able to write
+func canWrite(t *testing.T, args *testArgs) (bool, error) {
+	pktIO := getGDPParameter(t)
+	writeErr := writeTableEntry(args, t, pktIO, false)
+	if writeErr != nil {
+		if args.expectWrite == false {
+			t.Logf("Expected P4Client error: Client write permission (highID %d, lowID %d): want %v, got %v", args.highID, args.lowID, args.expectWrite, false)
+		}
+		return false, writeErr
+	}
+	return true, nil
+}
+
+// This function handles the values passed through the test case
+// and compares them to values obtained in canRead()/canWrite().
+func validateRWResp(t *testing.T, args *testArgs) bool {
+	returnReadVal := true
+	returnWriteVal := true
+	// Validate write permissions
+	writeOp, writeErr := canWrite(t, args)
+	if args.expectWrite != writeOp {
+		t.Errorf("Client write permission error: (highID %d, lowID %d): want %v, got %v", args.highID, args.lowID, args.expectWrite, writeOp)
+		if writeErr != nil {
+			t.Errorf("Write Error %v", writeErr)
+		}
+		returnWriteVal = false
+	}
+
+	// Validate read permissions
+	readOp, readErr := canRead(t, args)
+	if args.expectRead != readOp {
+		t.Errorf("Client read permission error: (highID %d, lowID %d): want %v, got %v", args.highID, args.lowID, args.expectRead, readOp)
+		t.Errorf("Read Error: %v", readErr)
+		returnReadVal = false
+	}
+	if returnWriteVal && returnReadVal {
+		return true
+	}
+	return false
+}
+
+func removeClient(handle *p4rt_client.P4RTClient) {
+	handle.StreamChannelDestroy(&streamName)
+	handle.ServerDisconnect()
+}
+
+// Test Zero client with 0 electionID
+func TestZeroMaster(t *testing.T) {
+	dut := ondatra.DUT(t, "dut")
+	configureDeviceId(t, dut)
+	configurePortId(t, dut)
+	test := testArgs{
+
+		desc:         pDesc,
+		lowID:        *ygot.Uint64(0),
+		highID:       highID,
+		handle:       clientConnection(t, dut),
+		deviceID:     deviceID,
+		expectPass:   false,
+		expectStatus: 0,
+	}
+	t.Run(test.desc, func(t *testing.T) {
```

Comment:
Do we expect to have >1 tests within this test case going forward? I guess the `t.Run` here is intended to just allow us to get a better failure report (since it seems like you couldn't just run this test case without the setup).